### PR TITLE
Transaction API

### PR DESCRIPTION
**Pull request by 0xFirekeeper**
Transactoooooooooooooor

### CODE REVIEW

File: Assets/Thirdweb/Core/Scripts/Transaction.cs [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+            return this;
+        }
+
+        public Transaction SetFrom(string from)
+        {
+            Input.From = from;
+            return this;
+        }
+
+        public Transaction SetGas(string gas)
```
Comment (joaquim-verges) [inline Assets/Thirdweb/Core/Scripts/Transaction.cs:71]: is this gas limit or gas price? should be explicit

File: Assets/Thirdweb/Core/Scripts/Transaction.cs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+            return gas.Value < defaultGas ? defaultGas : gas;
+        }
+
+        public async Task<GasCosts> EstimateGasCosts()
+        {
+            var gasLimit = await EstimateGasLimit();
+            var gasPrice = await GetGasPrice();
+            var gasCost = gasLimit * gasPrice;
+
+            return new GasCosts { ether = gasPrice.ToString().ToEth(18, false), wei = gasCost };
```
Comment (joaquim-verges) [inline Assets/Thirdweb/Core/Scripts/Transaction.cs:139]: niiice

File: Assets/Thirdweb/Core/Scripts/Transaction.cs [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+        }
+
+        public async Task<Transaction> EstimateAndSetGasAsync()
```
Comment (joaquim-verges) [inline Assets/Thirdweb/Core/Scripts/Transaction.cs:142]: when do I use this?

File: Assets/Thirdweb/Core/Scripts/Transaction.cs [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+        {
+            var gasBigInt = await EstimateGasLimit();
+            Input.Gas = gasBigInt.ToHexBigInteger();
+            return this;
+        }
+
+        public async Task<string> Send(bool? gasless = null)
```
Comment (joaquim-verges) [inline Assets/Thirdweb/Core/Scripts/Transaction.cs:149]: very nice

File: Assets/Thirdweb/Core/Scripts/Transaction.cs [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+
+                var postData = new RelayerRequest(request, signature, forwarderAddress);
+
+                using (UnityWebRequest req = UnityWebRequest.Post(relayerUrl, ""))
+                {
+                    byte[] bodyRaw = System.Text.Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(postData));
+                    req.uploadHandler = (UploadHandler)new UploadHandlerRaw(bodyRaw);
+                    req.downloadHandler = (DownloadHandler)new DownloadHandlerBuffer();
+                    req.SetRequestHeader("Content-Type", "application/json");
+                    await req.SendWebRequest();
```
Comment (joaquim-verges) [inline Assets/Thirdweb/Core/Scripts/Transaction.cs:231]: the gasless code is re-used with the transaction manager no? might be useful to consolidate

**Review by joaquim-verges — COMMENTED** [review]
excited for this!

File: Assets/Thirdweb/Core/Scripts/Transaction.cs (hunk 3 above)
Comment (0xFirekeeper) [reply Assets/Thirdweb/Core/Scripts/Transaction.cs:142]: If you manually change something like data I guess. I'll also probably use it for my typed transactions later.

**Pull request merged**